text/0000-safer-transmute.md

```diff
@@ -0,0 +1,1716 @@
+```rust
+/// Try to convert a `&T` into `&U`.
+///
+/// This produces `None` if the referent isn't appropriately
+/// aligned, as required by the destination type.
+pub fn try_cast_ref<'t, 'u, Scope, T, U>(src: &'t T) -> Option<&'u U>
+where
+    &'t T: TransmuteInto<&'u U, Scope, NeglectAlignment>,
+{
+    if (src as *const T as usize) % align_of::<U>() != 0 {
+        None
+    } else {
+        // Safe because we dynamically enforce the alignment
+        // requirement, whose static check we chose to neglect.
+        Some(unsafe { src.unsafe_transmute_into() })
+    }
+}
+```
+
+#### `NeglectValidity`
+By default, `TransmuteFrom` and `TransmuteInto`'s methods require that all instantiations of the source type are guaranteed to be valid instantiations of the destination type. This precludes transmutations which *might* be valid depending on the source value:
+```rust
+#[derive(PromiseTransmutableFrom, PromiseTransmutableInto)]
+#[repr(u8)]
+enum Bool {
+    True = 1,
+    False = 0,
+}
+
+/* ⚠️ This example intentionally does not compile. */
+let _ : Bool  = transmute!(some_u8_value); // Compile Error!
+```
+The `NeglectValidity` option disables this check.
+```rust
+pub struct NeglectValidity;
+
+impl TransmuteOptions for NeglectValidity {}
+```
+
+By using the `NeglectValidity` option, you are committing to ensure dynamically source value is a valid instance of the destination type. For instance:
+```rust
+#[derive(PromiseTransmutableFrom, PromiseTransmutableInto)]
+#[repr(u8)]
+enum Bool {
+    True = 1,
+    False = 0,
+}
+
+pub trait TryIntoBool<Scope>
+{
+    fn try_into_bool(self) -> Option<Bool>;
+}
+
+impl<T, Scope> TryIntoBool<Scope> for T
+where
+    T: TransmuteInto<u8>,
+    u8: TransmuteInto<Bool, Scope, NeglectValidity>
+{
+    fn try_into_bool(self) -> Option<Bool> {
+        let val: u8 = self.transmute_into();
+
+        if val > 1 {
+            None
+        } else {
+            // Safe, because we've first verified that
+            // `val` is a bit-valid instance of a boolean.
+            Some(unsafe {val.unsafe_transmute_into()})
+        }
+    }
+}
+```
+
+Even with `NeglectValidity`, the compiler will statically reject transmutations that cannot possibly be valid:
+```rust
+#[derive(PromiseTransmutableInto)]
+#[repr(C)] enum Foo { A = 24 }
+
+#[derive(PromiseTransmutableFrom)]
+#[repr(C)] enum Bar { Z = 42 }
+
+let _ = <Bar as TransmuteFrom<Foo, Here!(), NeglectValidity>::unsafe_transmute_from(Foo::N) // Compile error!
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Implementation Guidance
+Three items in this RFC require special compiler support:
+  - `Here!()`
+  - `Constructible` (a private implementation detail of `TransmuteFrom`)
+  - `TransmuteFrom`
+
+### Implementing `Here!()`
```

Comment:
Per @amluto's [comment](https://github.com/rust-lang/rfcs/pull/2981#issuecomment-706778920), the generated scope type should *probably* be private. This will cause rustc to produce an error if the scope type is publically re-exported via a type alias. "Leaking" a scope type isn't, on its own, problematic, but the consequences are kinda subtle.

The existing pub-in-priv lint isn't impossible to circumvent, but doing so requires a degree of intentionality that's appropriate to the situation.

Reply:
Addressed by https://github.com/rust-lang/rfcs/pull/2981/commits/a3250788d84dd3afbc5dede08ec9602fedd634a0.